processor/metricstransformprocessor/metrics_transform_processor_helper.go

```diff
@@ -0,0 +1,375 @@
+// Returns a map from keys to groups of timeseries and a map from keys to the corresponding label values
+func (mtp *metricsTransformProcessor) groupTimeseries(metric *metricspb.Metric, labelIdxs []int, newValue string, aggregatedValues []string) (map[string][]*metricspb.TimeSeries, map[string][]*metricspb.LabelValue) {
+	// key is a composite of the label values as a single string
+	// keyToTimeseriesMap groups timeseries by the label values
+	keyToTimeseriesMap := make(map[string][]*metricspb.TimeSeries)
+	// keyToLabelValuesMap maps the keys to the actual label values objects
+	keyToLabelValuesMap := make(map[string][]*metricspb.LabelValue)
+	for _, timeseries := range metric.Timeseries {
+		var composedValues string
+		var newLabelValues []*metricspb.LabelValue
+		if newValue == "" {
+			composedValues, newLabelValues = mtp.composeKeyWithSelectedLabels(labelIdxs, timeseries)
+		} else {
+			composedValues, newLabelValues = mtp.composeKeyWithNewValue(labelIdxs[0], timeseries, newValue, aggregatedValues)
+		}
+		// group the timeseries together if their values match
+		timeseriesGroup, ok := keyToTimeseriesMap[composedValues]
+		if ok {
+			keyToTimeseriesMap[composedValues] = append(timeseriesGroup, timeseries)
+		} else {
+			keyToTimeseriesMap[composedValues] = []*metricspb.TimeSeries{timeseries}
+			keyToLabelValuesMap[composedValues] = newLabelValues
+		}
+	}
+	return keyToTimeseriesMap, keyToLabelValuesMap
+}
+
+// composeKeyWithSelectedLabels composes the key for the timeseries based on the selected labels' values indicated by labelIdxs
+// Returns the key and a slice of the actual values used in this key
+func (mtp *metricsTransformProcessor) composeKeyWithSelectedLabels(labelIdxs []int, timeseries *metricspb.TimeSeries) (string, []*metricspb.LabelValue) {
+	var key string
+	// newLabelValues are the label values after aggregation, dropping the excluded ones
+	newLabelValues := make([]*metricspb.LabelValue, len(labelIdxs))
+	for i, vidx := range labelIdxs {
+		key += fmt.Sprintf("%v-", timeseries.LabelValues[vidx].Value)
+		newLabelValues[i] = timeseries.LabelValues[vidx]
+	}
+	return key, newLabelValues
+}
+
+// composeKeyWithNewValue composes the key for the timeseries with the aggregatedValues replaced by the newValue
+// Returns the key and a slice of the actual values used in this key
+func (mtp *metricsTransformProcessor) composeKeyWithNewValue(labelIdx int, timeseries *metricspb.TimeSeries, newValue string, aggregatedValues []string) (string, []*metricspb.LabelValue) {
+	aggregatredValuesSet := mtp.sliceToSet(aggregatedValues)
+	var key string
+	// newLabelValues are the label values after aggregation with the new value name
+	newLabelValues := make([]*metricspb.LabelValue, len(timeseries.LabelValues))
+	for i, labelValue := range timeseries.LabelValues {
+		if _, ok := aggregatredValuesSet[labelValue.Value]; labelIdx == i && ok {
+			key += fmt.Sprintf("%v-", newValue)
```

Comment:
I'm a little confused with that this function is doing - is it combining strings together with dashes? Is the idea to make a new value with all the different values combined with dashes?

Reply:
Yes, so here I am trying to use the values to construct a key to represent this combination of values. Based on my understanding, this would be the easiest way to construct a uniquely identifying key for a set of labelValues.